### PR TITLE
csi: volume ids are only unique per namespace

### PR DESCRIPTION
**Pull request by langmartin**
CSI volumes have had a namespace field, but it was not used correctly at the schema level to permit duplicate volume ids in different namespaces. This threads the namespace through code that addresses the volume by id. It's required by the volume ACL implementation (and would have presented as bug anyway) #7305

### CODE REVIEW

File: nomad/state/state_store.go [hunk 1]
```diff
@@ -1828,6 +1839,15 @@ func (s *StateStore) CSIVolumeDeregister(index uint64, ids []string) error {
 			return fmt.Errorf("volume not found: %s", id)
 		}
 
+		vol, ok := existing.(*structs.CSIVolume)
+		if !ok {
+			return fmt.Errorf("volume not found: %s", id)
+		}
+
+		if namespace != vol.Namespace {
```
Comment (tgross) [inline nomad/state/state_store.go]: Seems like we're missing some code in the body here?

File: nomad/csi_endpoint.go [hunk 2]
```diff
@@ -142,7 +142,7 @@ func (v *CSIVolume) List(args *structs.CSIVolumeListRequest, reply *structs.CSIV
 				}
 
 				// Filter on the request namespace to avoid ACL checks by volume
```
Comment (tgross) [inline nomad/csi_endpoint.go]: Do we still need this conditional if we're now querying the state store by namespace?

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -1759,6 +1757,19 @@ func (s *StateStore) CSIVolumesByNodeID(ws memdb.WatchSet, nodeID string) (memdb
...
+	if err != nil {
+		return nil, fmt.Errorf("volume lookup failed: %v", err)
+	}
+	ws.Add(iter.WatchCh())
+
+	return iter, nil
+}
+
 // CSIVolumes looks up the entire csi_volumes table
 func (s *StateStore) CSIVolumes(ws memdb.WatchSet) (memdb.ResultIterator, error) {
```
Comment (tgross) [inline nomad/state/state_store.go]: Does this have any callers now?

File: scheduler/feasible.go [hunk 4]
```diff
@@ -198,6 +199,10 @@ func NewCSIVolumeChecker(ctx Context) *CSIVolumeChecker {
 	}
 }
 
+func (c *CSIVolumeChecker) SetJob(job *structs.Job) {
```
Comment (tgross) [inline scheduler/feasible.go]: Isn't it `SetNamespace` (if we're comparing with `SetVolumes` below)? Or maybe `SetNamespaceFromJob` if you want to get really wordy. 😀

Comment (langmartin) [reply scheduler/feasible.go]: Well, this way it implements `ContextualIterator`, which is what I was going for. Looking again, it doesn't seem to matter.

File: nomad/state/state_store.go (hunk 3 above)
Comment (langmartin) [reply nomad/state/state_store.go]: Not anymore! I'll take it out

File: nomad/state/state_store.go (hunk 1 above)
Comment (langmartin) [reply nomad/state/state_store.go]: Ugh, thanks. This was the snippet of code from the ACL branch that got me started.

**Pull request merged**